### PR TITLE
[WIP] v2

### PR DESCRIPTION
**Pull request by Halleck45**
[Discussion is here](https://github.com/phpmetrics/PhpMetrics/issues/220)

[Live demo of HTML report](http://www.phpmetrics.org/report/v2/)

**Features**
- [x] improve HTML report
- [x] group metrics by category
- [x] way to get Graph and Tree metrics
- [x] analyzing more than one directory
- [x] parsing annotations

**Features in a wonderful word**
- [x] anti-pattern detection (following [DECOR](https://dibt.unimol.it/staff/fpalomba/documents/B1.pdf))
- [x] link between code and junit report

**New metrics**
- [x] ClassRank (Google's pageRank applyed to classes)
- [x] Git specific informations (optional)
- [ ] Third party (dependencies, known vulnerabilities) (optional)
- [x] Kan's defects
- [ ] npath complexity (optional)

**Depreciations**
- [x] removing the `--report-cli` option
- [x] removing the `--chart-bubbles` option
- [x] removing the `-level` option
- [x] removing the `--without-oop` option
- [x] removing the `--ignore-error` option
- [x] `--excluded-dirs` become `--exclude`
- [x] Removing plugin engine
- [x] removing the CLI detailled report

**Technical features**
- [x] improved parser (thanks to PhpParser)
- [x] work with graphs and nodes 
- [x] rewrite all metrics in order to work with graph and nodes
- [x] increase performance
- [ ] at least 100% of covered instructions for components and metrics
- [x] way to run application
- [x] simplify configuration and options
- [ ] improve website and documentation
- [x] remove some dependencies 
- [x] centralize packages (debian, brew, phparch...)

Thanks for your help !

### CODE REVIEW

File: src/Hal/Application/Application.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+        // config
+        $config = (new Parser())->parse($argv);
+        try {
+            (new Validator())->validate($config);
+        } catch (ConfigException $e) {
+
+            if ($config->has('help')) {
+                $output->writeln((new Validator())->help());
+                exit(1);
```
Comment (UFOMelkor) [inline src/Hal/Application/Application.php]: Why exiting with `1`?

File: src/Hal/Application/Application.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            }
+
+            if ($config->has('version')) {
+                $output->writeln(sprintf("PhpMetrics %s <http://phpmetrics.org>\nby Jean-François Lépine <https://twitter.com/Halleck45>", getVersion()));
+                exit(1);
```
Comment (UFOMelkor) [inline src/Hal/Application/Application.php]: Why exiting with `1`?

File: src/Hal/Application/Application.php (hunk 1 above)
Comment (Halleck45) [inline src/Hal/Application/Application.php]: Yes. I've just fixed it

File: src/Hal/Metric/Class_/ClassEnumVisitor.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
+<?php
+namespace Hal\Metric\Class_;
+
+use Hal\Component\Reflected\Method;
```
Comment (UFOMelkor) [inline src/Hal/Metric/Class_/ClassEnumVisitor.php]: Unused import (and unknown class)

File: src/Hal/Metric/Class_/Component/MaintainabilityIndexVisitor.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * @inheritdoc
+     */
+    public function leaveNode(Node $node)
+    {
+        if ($node instanceof Stmt\Class_) {
+
+            if ($node instanceof Stmt\Class_) {
+                $classOrFunction = $this->metrics->get($node->namespacedName->toString());
+            } else {
```
Comment (UFOMelkor) [inline src/Hal/Metric/Class_/Component/MaintainabilityIndexVisitor.php]: `else` would never happen

**Pull request merged**